### DISCUSSION TITLE
Is there a way to bind project as multiple services?

### DISCUSSION THREAD

**Discussion by En3Tho** [microsoft/aspire#959] [Q&A] [answered]
I have a host project that actually hosts 2 services, "apiservice" and "apiservicev2" for example. How can I specify an additional service name for a project that I add to the `DistributedApplicationBuilder`?

Basically what I want to do is something like that:

```csharp
var apiservice = builder.AddProject<Projects.ApiService>("apiservice", "apiservicev2"); // implements both at once and references get injected automatically when referenced
var apiservice2 = builder.AddProject<Projects.ApiService2>("apiservice2")
    .WithReference(apiservice); // knows about both "apiservice" and "apiservicev2"
```

or
```csharp
var apiservice = builder.AddProject<Projects.ApiService>("apiservice"); // default name is apiservice
var apiservice2 = builder.AddProject<Projects.ApiService2>("apiservice2")
    .WithReference(apiservice)
    .WithReference(apiservice, name: "apiservicev2"); // explicit custom name
```

I've tried some things but they did not work. It is quite possible that I've missed something in the doc.

**Comment by davidfowl** [comment]
> I have a host project that actually hosts 2 services, "apiservice" and "apiservicev2" for example.

Can you explain some more? Do you want to declare 2 services or just refer to the same service by a different name? If so can you explain why?

**Reply by En3Tho** [reply]
I want to refer to the same "host" by multiple different names ("services"). Basically my services are modules that expose both client registration and server registation and can be hosted both as part of modular monolith or as a dedicated microservice. The idea is that instead of using HttpClient or whatever client directly, I expose "IMyBlablaService" interface with assossiated contracts (dtos and such) and end code always knows only about that. It doesn't know where service is hosted or what protocol is used etc. I'm experimenting with Aspire as a unified dev host (which I'm loving so far, it's great!).

Imagine this:
```
ServiceA : IServiceA
  => exposes ServiceA.Contracts (shared lib with interface, dtos and stuff like that)
  => exposes ServiceA.Client ("http://serviceA")
    services.AddServiceAClient()
  => exposes ServiceA.Server
    services.AddServiceAServer()
    app.MapServiceA()

ServiceB : ... same as above but maybe a grpc and not http, the point is that only IServiceB interface is consumed in the end

Project1
   => hosts ServiceAServer
   => hosts ServiceBServer
Project2
  => consumes ServiceAClient
  => consumes ServiceBClient
```

So I might host "serviceA" and "serviceB" in the same host "Project1" and I need a way for "Project2" to be able to discover both "serviceA" and "serviceB".

I hope this clarifies things a bit.

**Reply by En3Tho** [reply]
I guess you can think of it as "alias" to the project in terms of `Aspire`

**Reply by davidfowl** [reply, marked as answer]
There's no first-class support for doing this today but I think you can implement it. This won't work for deployment though, I'll file an issue:

```C#
static class WithReferenceAlias
{
    public static IResourceBuilder<ProjectResource> WithReference(
        this IResourceBuilder<ProjectResource> destination,
        IResourceBuilder<ProjectResource> source,
        string[] aliases)
    {
        static bool ContainsAmbiguousEndpoints(IEnumerable<AllocatedEndpointAnnotation> endpoints)
        {
            // An ambiguous endpoint is where any scheme (
            return endpoints.GroupBy(e => e.UriScheme).Any(g => g.Count() > 1);
        }

        return destination.WithEnvironment(context =>
        {
            if (!source.Resource.TryGetAllocatedEndPoints(out var allocatedEndPoints))
            {
                return;
            }

            var containsAmbiguousEndpoints = ContainsAmbiguousEndpoints(allocatedEndPoints);

            aliases = [source.Resource.Name, .. aliases];

            foreach (var name in aliases.Distinct())
            {
                var i = 0;
                foreach (var allocatedEndPoint in allocatedEndPoints)
                {
                    var bindingNameQualifiedUriStringKey = $"services__{name}__{i++}";
                    context.EnvironmentVariables[bindingNameQualifiedUriStringKey] = allocatedEndPoint.BindingNameQualifiedUriString;

                    if (!containsAmbiguousEndpoints)
                    {
                        var uriStringKey = $"services__{name}__{i++}";
                        context.EnvironmentVariables[uriStringKey] = allocatedEndPoint.UriString;
                    }
                }
            }
        });
    }
}
```

You should be able to do this:

```C#
var apiservice = builder.AddProject<Projects.ApiService>("apiservice"); // default name is apiservice
var apiservice2 = builder.AddProject<Projects.ApiService2>("apiservice2")
    .WithReference(apiservice)
    .WithReference(apiservice, ["apiservicev2"]); 
```

**Reply by En3Tho** [reply]
Thank you. I've been looking at the same code too. What do you personally think about this kind of project structure?

**Reply by davidfowl** [reply]
The only issue you will have with this pattern is that you will need to do something to distinguish between those services when you call them (if they are on the same port). If you expose service A and B from `apiservice`, then you can't really refer to them as "http://a" and "http://b", they'll be on the same port and you'll need to handle the routing somehow.

**Reply by davidfowl** [reply]
How are you going to handle that?

**Reply by davidfowl** [reply]
To makes this work you might need to include the path in there:

// Logical names
serviceA
serviceB

// Modular monolith

http://services/A -> http://serviceA
http://services/B -> http://serviceB

// Microservice

http://serviceA
http://serviceB

@ReubenBond Can service discovery handle the path as input for http requests?

**Reply by En3Tho** [reply]
Do you mean when services are deployed in production? Example you gave simple mapped "http://a/" and "http://b/" to the same address "http://localhost:63558" and that worked without issues (well, locally). I belive in production there would be a similar mechanism. I'm not sure about mixing http and grpc like that but this is my next objective.

**Reply by En3Tho** [reply]
Oh, when deployed in modular monolith services call each other w/o http. They just simply call `IBlabla` interface and that executes (a)synchronously inside the monolith. Monolight mode never calls `AddClient`, only `AddServer`.

**Reply by davidfowl** [reply]
What's the situation when you want to refer to a single service as different names?

**Reply by En3Tho** [reply]
Well, the general idea is that by hosting multiple logical services in one place multiple things can be shared: GC, connections, threads. Calling services in memory is obviously faster etc. But sometimes service might need additional resources and thus moved to dedicated hosts/pods.

So when a monolith implements multiple services you end up with same base host url for both.

This obviously only works for "inner-circle" services that do not require complex auth or something like that. They just expose a set of contracts and 2 ways of using them (server implementation and client implementation for DI).

**Reply by ReubenBond** [reply]
> Can service discovery handle the path as input for http requests?

In theory, yes, but the way it's implemented right now it doesn't. Partly for efficiency: we key only on the authority because the path is likely to vary arbitrarily (eg thousands of entries) so if we didn't strip the path & query, we may end up with an arbitrarily large cache of resolvers polling / waiting for updates. If we gave resolvers a way to signal which parts of the input were relevant, this might be feasible. Otherwise, I think we end up with _client-side YARP_, which feels to me like a separate library.
